chanzuckerberg/idseq-web
# Automatic pipeline restart

# Description

This change intends to reduce the on-call noise by automatically restarting the pipeline whenever a pipeline run fails for some random reason. The rspec results below describe this behavior in detail.

```
PipelineRun
  #update_job_status
    when all stages complete successfully
      changes status to finalized
    when a stage completes with an error
      and it is a known user error
        changes status to failed
        does not send error to airbrake
        does not enqueue new pipeline run
      and it is not a known user error
        and an automatic restart is allowed
          changes status to failed
          sends error to airbrake
          enqueues new pipeline run
        and an automatic restart is not allowed
          changes status to failed
          sends error to airbrake
          does not enqueue new pipeline run
  #automatic_restart_allowed?
    when user is an admin
      should be falsy
    when user is not an admin
      and sample has no previous pipeline runs with the same pipeline version
        should be truthy
      and sample has previous pipeline runs with the same pipeline version
        and they all succeeded
          should be truthy
        and at least one of them failed
          should be falsy
  #report_failed_pipeline_run_stage
    sends metric to datadog
    when send_to_airbrake is true
      sends error to airbrake and log error
    when send_to_airbrake is false
      do not send error to airbrake and log warn

Finished in 0.81819 seconds (files took 7.13 seconds to load)
17 examples, 0 failures
```

## Type of change


# Test and Reproduce
Outline the steps to test or reproduce the PR here.

1. Upload a sample
2. Terminate the job in AWS console to trigger a failure
3. Wait a few minutes and a new pipeline run should kick off. No error messages so far.
4. Terminate the job in AWS console for this second pipeline run.
5. Pipeline shouldn't restart again. We should see an error being sent to Airbrake.

## Impacted Areas in Application
List general components of the application that this PR will affect:


Specific components 

* [X] pipeline monitor


# Checklist:
